src/main/java/com/google/devtools/build/lib/rules/cpp/ClangHeaderMap.java

```diff
@@ -0,0 +1,325 @@
+package com.google.devtools.build.lib.rules.cpp;
+
+import java.io.File;
+import java.io.FileOutputStream;
+
+import java.util.Map;
+import java.util.HashMap;
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.List;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.ByteOrder;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.channels.FileChannel;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import static java.lang.Math.max;
+
+public final class ClangHeaderMap {
+  // Logical representation of a bucket.
+  // The actual data is stored in the string pool.
+  private class HMapBucket {
+    String key;
+    String prefix;
+    String suffix;
+
+    HMapBucket(String key, String prefix, String suffix) {
+      this.key = key;
+      this.prefix = prefix;
+      this.suffix = suffix;
+    }
+  }
+
+  private static final int HEADER_MAGIC = ('h' << 24) | ('m' << 16) | ('a' << 8) | 'p';
+  private static final short HEADER_VERSION = 1;
+  private static final short HEADER_RESERVED = 0;
+  private static final int EMPTY_BUCKET_KEY = 0;
+
+  private static final int HEADER_SIZE = 24;
+  private static final int BUCKET_SIZE = 12;
+
+  private static final int INT_SIZE = Integer.SIZE/8;
+
+  // Data stored in accordance to Clang's lexer types
```

Comment:
Make part of the javadoc below, separate with blank line from the code, and wrap code snippet in

 * \<pre\>{@code
 * Your code here.
 * }\</pre\>

Reply:
Thanks for the suggestion, will fix